service/node.go

```diff
@@ -273,17 +282,33 @@ func (n *nodeService) UpdateReport(namespace, name string, report specV1.Report)
 }
 
 // UpdateDesire Update Desire
-func (n *nodeService) UpdateDesire(namespace, name string, desire specV1.Desire) (*models.Shadow, error) {
-	shadow, err := n.shadow.Get(namespace, name)
-	if err != nil {
-		return nil, err
-	}
+// Parameter f can be RefreshNodeDesireByApp or DeleteNodeDesireByApp
+func (n *nodeService) UpdateDesire(shadow *models.Shadow, app *specV1.Application, f func(*models.Shadow, *specV1.Application)) (*models.Shadow, error) {
+	// Retry times
+	var count = 0
+	for {
+		newShadow, err := n.shadow.Get(shadow.Namespace, shadow.Name)
+		if err != nil {
+			return nil, err
+		}
 
-	if shadow == nil {
-		return n.createShadow(namespace, name, desire, nil)
-	}
+		if newShadow == nil {
+			f(shadow, app)
+			return n.createShadow(shadow.Namespace, shadow.Name, shadow.Desire, nil)
+		}
 
-	return n.updateDesire(shadow, desire)
+		// Refresh desire in shadow by app
+		f(newShadow, app)
+		updatedShadow, err := n.shadow.UpdateDesire(newShadow)
+		if err == nil || err.Error() != "invalid version" {
```

Comment:
add return update result（success or failed）and check the update result in the if statement